lnwallet/btcwallet/btcwallet.go

```diff
@@ -1199,33 +1200,98 @@ func (b *BtcWallet) ListUnspentWitness(minConfs, maxConfs int32,
 // will be returned.
 func (b *BtcWallet) PublishTransaction(tx *wire.MsgTx, label string) error {
-	if err := b.wallet.PublishTransaction(tx, label); err != nil {
+	// handleErr is a helper closure that handles the error from
+	// PublishTransaction and TestMempoolAccept.
+	handleErr := func(err error) error {
+		var (
+			doubleSpendErr      *base.ErrDoubleSpend
+			replacementErr      *base.ErrReplacement
+			mempoolFeeErr       *base.ErrMempoolFee
+			inMempoolErr        *base.ErrInMempool
+			alreadyConfirmedErr *base.ErrAlreadyConfirmed
+		)
+
 		// If we failed to publish the transaction, check whether we
 		// got an error of known type.
-		switch err.(type) {
+		switch {
 		// If the wallet reports a double spend, convert it to our
 		// internal ErrDoubleSpend and return.
-		case *base.ErrDoubleSpend:
+		case errors.As(err, &doubleSpendErr):
 			return lnwallet.ErrDoubleSpend
 
 		// If the wallet reports a replacement error, return
 		// ErrDoubleSpend, as we currently are never attempting to
 		// replace transactions.
-		case *base.ErrReplacement:
+		case errors.As(err, &replacementErr):
 			return lnwallet.ErrDoubleSpend
 
-		// If the wallet reports that fee requirements for accepting the
-		// tx into mempool are not met, convert it to our internal
+		// If the wallet reports that fee requirements for accepting
+		// the tx into mempool are not met, convert it to our internal
 		// ErrMempoolFee and return.
-		case *base.ErrMempoolFee:
+		case errors.As(err, &mempoolFeeErr):
 			return fmt.Errorf("%w: %v", lnwallet.ErrMempoolFee,
 				err.Error())
 
+		//nolint:lll
+		// These two errors are ignored inside `PublishTransaction`:
+		// https://github.com/btcsuite/btcwallet/blob/master/wallet/wallet.go#L3763
+		// To keep our current behavior, we need to ignore the same
+		// errors returned from TestMempoolAccept.
+		//
+		// TODO(yy): since `LightningWallet.PublishTransaction` always
+		// publish the same tx twice, we'd always get ErrInMempool. We
+		// should instead create a new rebroadcaster that monitors the
+		// mempool, and only rebroadcast when the tx is evicted. This
+		// way we don't need to broadcast twice, and can instead return
+		// these errors here.
+		case errors.As(err, &inMempoolErr),
+			errors.As(err, &alreadyConfirmedErr):
+
+			return nil
+
 		default:
 			return err
 		}
 	}
-	return nil
+
+	// For neutrino backend there's no mempool, so we return early by
+	// publishing the transaction.
+	if b.chain.BackEnd() == "neutrino" {
+		err := b.wallet.PublishTransaction(tx, label)
+		return handleErr(err)
+	}
+
+	// For non-neutrino nodes, we will first check whether the transaction
+	// can be accepted by the mempool.
+	// Use a max feerate of 0 means the default value will be used when
+	// testing mempool acceptance. The default max feerate is 0.10 BTC/kvb,
+	// or 10,000 sat/vb.
+	results, err := b.chain.TestMempoolAccept([]*wire.MsgTx{tx}, 0)
```

Comment:
What's the fee rate actually mean here? Don't we want to use the fee rate of the transaction itself? Though I guess at this point, the function being executed doesn't have access to the input amount, so we can't actually compute what the effective fee is....

Reply:
Also looking at the API now, I wonder if we should have a like `PackagePublishTx` method. We'd use this when doing things like the `BumpFee` fee bump, or areas where the sweeper _knows_ it's actually doing a fee bump. 

I also wonder if API wise, there's a way to have `testmempoolaccept` actually compute the effective fee rate for us? So like as part of the return body.

Reply:
> What's the fee rate actually mean here?

Do you mean the param `maxfeerate`? I think it's used to make sure the calculated fee rate of this tx doesn't exceed this value.

> I wonder if we should have a like PackagePublishTx method.

This will be handled specifically in the sweeper - the idea is to use `testmempoolaccept` to check for RBF compliance before actually publishing a sweeping tx, thus we don't need to reimplement the RBF validation logic in `lnd` again.

> I also wonder if API wise, there's a way to have testmempoolaccept actually compute the effective fee rate for us? So like as part of the return body.

Yeah I think so, tho the effective feerate is mostly useful for CPFPing using anchors?

Reply:
> I also wonder if API wise, there's a way to have testmempoolaccept actually compute the effective fee rate for us? So like as part of the return body.

On the Core side ideally. Like it rejects a transaction that can't replace another, and tells the effective fee rate and package size. I think right now it just says accept or deny, w/o any other details.

Reply:
> Yeah I think so, tho the effective feerate is mostly useful for CPFPing using anchors?

Yeah